File: backend/app/schemas/gs1.py
"""GS1 Barcode + Label Printing — Pydantic v2 schemas."""
from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, field_validator


# ── Shared base ───────────────────────────────────────────────────────────────

class _Base(BaseModel):
    model_config = ConfigDict(from_attributes=True)


# ── GS1 Company Config ────────────────────────────────────────────────────────

class GS1CompanyConfigCreate(_Base):
    company_name: str
    gs1_company_prefix: str
    country_code: Optional[str] = None
    extension_digit: int = 0
    notes: Optional[str] = None


class GS1CompanyConfigOut(_Base):
    id: uuid.UUID
    company_name: str
    gs1_company_prefix: str
    country_code: Optional[str]
    extension_digit: int
    sscc_serial_counter: int
    is_active: bool
    notes: Optional[str]
    created_at: datetime


# ── Product GS1 Config ────────────────────────────────────────────────────────

class ProductGS1ConfigCreate(_Base):
    product_id: uuid.UUID
    company_config_id: Optional[uuid.UUID] = None
    gtin: Optional[str] = None
    barcode_type: str = "EAN13"
    packaging_level: str = "UNIT"
    label_template_id: Optional[uuid.UUID] = None
    gs1_company_prefix: Optional[str] = None
    requires_serialization: bool = False
    requires_lot_tracking: bool = True
    requires_expiry_tracking: bool = True
    units_per_inner_pack: Optional[int] = None
    inner_packs_per_carton: Optional[int] = None
    cartons_per_pallet: Optional[int] = None
    carton_gtin: Optional[str] = None
    pallet_gtin: Optional[str] = None
    notes: Optional[str] = None


class ProductGS1ConfigUpdate(_Base):
    gtin: Optional[str] = None
    barcode_type: Optional[str] = None
    packaging_level: Optional[str] = None
    label_template_id: Optional[uuid.UUID] = None
    requires_serialization: Optional[bool] = None
    requires_lot_tracking: Optional[bool] = None
    requires_expiry_tracking: Optional[bool] = None
    units_per_inner_pack: Optional[int] = None
    inner_packs_per_carton: Optional[int] = None
    cartons_per_pallet: Optional[int] = None
    carton_gtin: Optional[str] = None
    pallet_gtin: Optional[str] = None
    notes: Optional[str] = None


class ProductGS1ConfigOut(_Base):
    id: uuid.UUID
    product_id: uuid.UUID
    company_config_id: Optional[uuid.UUID]
    gtin: Optional[str]
    barcode_type: str
    packaging_level: str
    label_template_id: Optional[uuid.UUID]
    gs1_company_prefix: Optional[str]
    requires_serialization: bool
    requires_lot_tracking: bool
    requires_expiry_tracking: bool
    units_per_inner_pack: Optional[int]
    inner_packs_per_carton: Optional[int]
    cartons_per_pallet: Optional[int]
    carton_gtin: Optional[str]
    pallet_gtin: Optional[str]
    is_active: bool
    notes: Optional[str]
    product_name: Optional[str] = None
    created_at: datetime


# ── Barcode Generation ────────────────────────────────────────────────────────

class BarcodeGenerateRequest(_Base):
    product_id: Optional[uuid.UUID] = None
    gtin: Optional[str] = None
    lot_number: Optional[str] = None
    expiry_date: Optional[date] = None
    production_date: Optional[date] = None
    serial_number: Optional[str] = None
    barcode_type: str = "GS1_128"
    quantity: Optional[Decimal] = None
    lot_id: Optional[uuid.UUID] = None
    packaging_level: Optional[str] = None
    save_record: bool = True


class BarcodeGenerateResponse(_Base):
    gs1_ai_string: str
    gs1_raw_string: str
    qr_payload: str
    barcode_image_b64: Optional[str] = None
    qr_image_b64: Optional[str] = None
    check_digit_valid: bool = True
    record_id: Optional[uuid.UUID] = None
    gtin: str
    lot_number: Optional[str]
    expiry_date: Optional[date]
    production_date: Optional[date]


# ── GS1 String Parsing ────────────────────────────────────────────────────────

class GS1DecodeRequest(_Base):
    gs1_string: str


class GS1DecodeResponse(_Base):
    segments: dict[str, str]
    gtin: Optional[str] = None
    lot_number: Optional[str] = None
    expiry_date: Optional[str] = None
    production_date: Optional[str] = None
    serial_number: Optional[str] = None
    sscc: Optional[str] = None
    is_valid: bool
    validation_errors: List[str] = []
    matched_product: Optional[dict] = None
    matched_lot: Optional[dict] = None


# ── Lot Barcode Record ────────────────────────────────────────────────────────

class LotBarcodeOut(_Base):
    id: uuid.UUID
    product_config_id: uuid.UUID
    lot_id: Optional[uuid.UUID]
    gtin: str
    lot_number: str
    production_date: Optional[date]
    expiry_date: Optional[date]
    serial_number: Optional[str]
    barcode_type: str
    gs1_ai_string: str
    qr_payload: Optional[str]
    barcode_image_b64: Optional[str]
    qr_image_b64: Optional[str]
    is_printed: bool
    printed_at: Optional[datetime]
    quantity: Optional[Decimal]
    packaging_level: Optional[str]
    created_at: datetime


# ── SSCC / Pallet ─────────────────────────────────────────────────────────────

class SSCCGenerateRequest(_Base):
    company_config_id: Optional[uuid.UUID] = None
    pallet_id: Optional[str] = None
    warehouse_location: Optional[str] = None
    notes: Optional[str] = None


class SSCCPalletOut(_Base):
    id: uuid.UUID
    sscc_code: str
    extension_digit: int
    company_prefix: str
    serial_reference: str
    check_digit: int
    pallet_id: Optional[str]
    status: str
    warehouse_location: Optional[str]
    shipment_reference: Optional[str]
    sscc_image_b64: Optional[str]
    label_printed: bool
    label_printed_at: Optional[datetime]
    notes: Optional[str]
    lot_count: int = 0
    created_at: datetime


class SSCCAddLotRequest(_Base):
    lot_barcode_id: uuid.UUID
    quantity: Optional[Decimal] = None
    packaging_level: Optional[str] = None


# ── Label Template ─────────────────────────────────────────────────────────────

class LabelTemplateCreate(_Base):
    name: str
    description: Optional[str] = None
    packaging_level: Optional[str] = None
    width_mm: Optional[Decimal] = None
    height_mm: Optional[Decimal] = None
    fields: Optional[dict] = None
    html_template: Optional[str] = None
    css_styles: Optional[str] = None
    is_default: bool = False
    notes: Optional[str] = None


class LabelTemplateUpdate(_Base):
    name: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    width_mm: Optional[Decimal] = None
    height_mm: Optional[Decimal] = None
    fields: Optional[dict] = None
    html_template: Optional[str] = None
    css_styles: Optional[str] = None
    is_default: Optional[bool] = None
    notes: Optional[str] = None


class LabelTemplateOut(_Base):
    id: uuid.UUID
    name: str
    description: Optional[str]
    packaging_level: Optional[str]
    template_version: int
    status: str
    width_mm: Optional[Decimal]
    height_mm: Optional[Decimal]
    fields: Optional[dict]
    html_template: Optional[str]
    css_styles: Optional[str]
    is_default: bool
    is_active: bool
    notes: Optional[str]
    created_at: datetime


# ── Print Job ─────────────────────────────────────────────────────────────────

class PrintJobCreate(_Base):
    trigger: str = "MANUAL"
    template_id: Optional[uuid.UUID] = None
    printer_name: Optional[str] = None
    printed_by: Optional[str] = None
    notes: Optional[str] = None
    items: List[dict] = []


class PrintJobOut(_Base):
    id: uuid.UUID
    job_no: str
    trigger: str
    template_id: Optional[uuid.UUID]
    status: str
    printed_by: Optional[str]
    printed_at: Optional[datetime]
    quantity: int
    printer_name: Optional[str]
    notes: Optional[str]
    item_count: int = 0
    created_at: datetime


# ── AI Recommendations ────────────────────────────────────────────────────────

class GS1AIRecOut(_Base):
    id: uuid.UUID
    agent_type: str
    title: str
    description: str
    affected_entity_type: Optional[str]
    affected_entity_id: Optional[str]
    recommendation: str
    severity: str
    status: str
    reviewed_by: Optional[str]
    reviewed_at: Optional[datetime]
    agent_metadata: Optional[dict]
    created_at: datetime


class GS1AIRecReview(_Base):
    status: str
    reviewed_by: Optional[str] = None


# ── Dashboard ─────────────────────────────────────────────────────────────────

class GS1DashboardOut(_Base):
    total_products_configured: int
    total_gtins: int
    total_barcodes_generated: int
    total_labels_printed: int
    total_sscc_pallets: int
    active_sscc_pallets: int
    total_print_jobs: int
    pending_print_jobs: int
    total_label_templates: int
    ai_recommendations_pending: int
    recent_barcodes: List[LotBarcodeOut]
    recent_sscc: List[SSCCPalletOut]


# ── Report ────────────────────────────────────────────────────────────────────

class LabelPrintHistoryOut(_Base):
    job_no: str
    trigger: str
    status: str
    printed_by: Optional[str]
    printed_at: Optional[datetime]
    quantity: int
    template_name: Optional[str]
    created_at: datetime


class SSCCTrackingOut(_Base):
    sscc_code: str
    pallet_id: Optional[str]
    status: str
    warehouse_location: Optional[str]
    shipment_reference: Optional[str]
    lot_count: int
    created_at: datetime


class PackagingHierarchyOut(_Base):
    gtin: str
    product_name: str
    packaging_level: str
    units_per_inner_pack: Optional[int]
    inner_packs_per_carton: Optional[int]
    cartons_per_pallet: Optional[int]
    total_units_per_pallet: Optional[int]

File: frontend/src/lib/gs1.ts
// GS1 Barcode + Label Printing System — types & API client

const BASE = `${process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000"}/api/v1/gs1`;

// ── Enums / Types ─────────────────────────────────────────────────────────────

export type BarcodeType = "EAN13" | "GS1_128" | "QR_CODE" | "CODE128" | "GS1_DATAMATRIX";
export type PackagingLevel = "UNIT" | "INNER_PACK" | "CARTON" | "PALLET";
export type PrintJobStatus = "PENDING" | "PRINTING" | "COMPLETED" | "FAILED" | "CANCELLED";
export type SSCCStatus = "ACTIVE" | "SHIPPED" | "RECEIVED" | "CANCELLED";
export type LabelTemplateStatus = "DRAFT" | "ACTIVE" | "DEPRECATED";
export type PrintTrigger = "MANUAL" | "PRODUCTION_COMPLETE" | "PACKAGING_COMPLETE" | "PALLETIZATION" | "SHIPMENT" | "QC_RELEASE";
export type GS1AIAgentType = "LABEL_VALIDATOR" | "PACKAGING_OPTIMIZER";
export type GS1AIRecStatus = "PENDING" | "ACCEPTED" | "REJECTED" | "APPLIED";

// ── Domain Objects ────────────────────────────────────────────────────────────

export interface GS1CompanyConfig {
  id: string;
  company_name: string;
  gs1_company_prefix: string;
  country_code: string | null;
  extension_digit: number;
  sscc_serial_counter: number;
  is_active: boolean;
  notes: string | null;
  created_at: string;
}

export interface ProductGS1Config {
  id: string;
  product_id: string;
  company_config_id: string | null;
  gtin: string | null;
  barcode_type: BarcodeType;
  packaging_level: PackagingLevel;
  label_template_id: string | null;
  gs1_company_prefix: string | null;
  requires_serialization: boolean;
  requires_lot_tracking: boolean;
  requires_expiry_tracking: boolean;
  units_per_inner_pack: number | null;
  inner_packs_per_carton: number | null;
  cartons_per_pallet: number | null;
  carton_gtin: string | null;
  pallet_gtin: string | null;
  is_active: boolean;
  notes: string | null;
  product_name: string | null;
  created_at: string;
}

export interface LotBarcodeRecord {
  id: string;
  product_config_id: string;
  lot_id: string | null;
  gtin: string;
  lot_number: string;
  production_date: string | null;
  expiry_date: string | null;
  serial_number: string | null;
  barcode_type: BarcodeType;
  gs1_ai_string: string;
  qr_payload: string | null;
  barcode_image_b64: string | null;
  qr_image_b64: string | null;
  is_printed: boolean;
  printed_at: string | null;
  quantity: string | null;
  packaging_level: PackagingLevel | null;
  created_at: string;
}

export interface BarcodeGenerateRequest {
  product_id?: string;
  gtin?: string;
  lot_number?: string;
  expiry_date?: string;
  production_date?: string;
  serial_number?: string;
  barcode_type?: BarcodeType;
  quantity?: number;
  lot_id?: string;
  packaging_level?: PackagingLevel;
  save_record?: boolean;
}

export interface BarcodeGenerateResponse {
  gs1_ai_string: string;
  gs1_raw_string: string;
  qr_payload: string;
  barcode_image_b64: string | null;
  qr_image_b64: string | null;
  check_digit_valid: boolean;
  record_id: string | null;
  gtin: string;
  lot_number: string | null;
  expiry_date: string | null;
  production_date: string | null;
}

export interface GS1DecodeResponse {
  segments: Record<string, string>;
  gtin: string | null;
  lot_number: string | null;
  expiry_date: string | null;
  production_date: string | null;
  serial_number: string | null;
  sscc: string | null;
  is_valid: boolean;
  validation_errors: string[];
  matched_product: Record<string, string> | null;
  matched_lot: Record<string, string> | null;
}

export interface SSCCPallet {
  id: string;
  sscc_code: string;
  extension_digit: number;
  company_prefix: string;
  serial_reference: string;
  check_digit: number;
  pallet_id: string | null;
  status: SSCCStatus;
  warehouse_location: string | null;
  shipment_reference: string | null;
  sscc_image_b64: string | null;
  label_printed: boolean;
  label_printed_at: string | null;
  notes: string | null;
  lot_count: number;
  created_at: string;
}

export interface LabelTemplate {
  id: string;
  name: string;
  description: string | null;
  packaging_level: PackagingLevel | null;
  template_version: number;
  status: LabelTemplateStatus;
  width_mm: string | null;
  height_mm: string | null;
  fields: Record<string, unknown> | null;
  html_template: string | null;
  css_styles: string | null;
  is_default: boolean;
  is_active: boolean;
  notes: string | null;
  created_at: string;
}

export interface PrintJob {
  id: string;
  job_no: string;
  trigger: PrintTrigger;
  template_id: string | null;
  status: PrintJobStatus;
  printed_by: string | null;
  printed_at: string | null;
  quantity: number;
  printer_name: string | null;
  notes: string | null;
  item_count: number;
  created_at: string;
}

export interface GS1AIRecommendation {
  id: string;
  agent_type: GS1AIAgentType;
  title: string;
  description: string;
  affected_entity_type: string | null;
  affected_entity_id: string | null;
  recommendation: string;
  severity: string;
  status: GS1AIRecStatus;
  reviewed_by: string | null;
  reviewed_at: string | null;
  agent_metadata: Record<string, unknown> | null;
  created_at: string;
}

export interface GS1Dashboard {
  total_products_configured: number;
  total_gtins: number;
  total_barcodes_generated: number;
  total_labels_printed: number;
  total_sscc_pallets: number;
  active_sscc_pallets: number;
  total_print_jobs: number;
  pending_print_jobs: number;
  total_label_templates: number;
  ai_recommendations_pending: number;
  recent_barcodes: LotBarcodeRecord[];
  recent_sscc: SSCCPallet[];
}

// ── API Client ────────────────────────────────────────────────────────────────

function authHeader(): Record<string, string> {
  const token = localStorage.getItem("access_token");
  return token ? { Authorization: `Bearer ${token}` } : {};
}

async function req<T>(method: string, path: string, body?: unknown): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    method,
    headers: { "Content-Type": "application/json", ...authHeader() },
    body: body !== undefined ? JSON.stringify(body) : undefined,
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error(err.detail || res.statusText);
  }
  return res.json();
}

export const gs1Api = {
  // Dashboard
  getDashboard: () => req<GS1Dashboard>("GET", "/dashboard"),

  // Company config
  listCompanyConfigs: () => req<GS1CompanyConfig[]>("GET", "/config"),
  createCompanyConfig: (d: Partial<GS1CompanyConfig>) => req<GS1CompanyConfig>("POST", "/config", d),

  // Product GS1 config
  listProductConfigs: () => req<ProductGS1Config[]>("GET", "/products"),
  getProductConfig: (id: string) => req<ProductGS1Config>("GET", `/products/${id}`),
  getConfigByProduct: (productId: string) => req<ProductGS1Config>("GET", `/products/by-product/${productId}`),
  createProductConfig: (d: Partial<ProductGS1Config>) => req<ProductGS1Config>("POST", "/products", d),
  updateProductConfig: (id: string, d: Partial<ProductGS1Config>) => req<ProductGS1Config>("PATCH", `/products/${id}`, d),

  // Barcode generation
  generateBarcode: (d: BarcodeGenerateRequest) => req<BarcodeGenerateResponse>("POST", "/barcode/generate", d),
  listBarcodes: (params?: { product_config_id?: string; lot_id?: string; limit?: number }) => {
    const q = new URLSearchParams();
    if (params?.product_config_id) q.set("product_config_id", params.product_config_id);
    if (params?.lot_id) q.set("lot_id", params.lot_id);
    if (params?.limit) q.set("limit", String(params.limit));
    return req<LotBarcodeRecord[]>("GET", `/barcode?${q}`);
  },
  getBarcode: (id: string) => req<LotBarcodeRecord>("GET", `/barcode/${id}`),

  // Scan / decode
  decodeGS1: (gs1_string: string) => req<GS1DecodeResponse>("POST", "/scan/decode", { gs1_string }),

  // SSCC
  generateSSCC: (d?: { company_config_id?: string; pallet_id?: string; warehouse_location?: string; notes?: string }) =>
    req<SSCCPallet>("POST", "/sscc/generate", d || {}),
  listSSCC: (params?: { status?: string; limit?: number }) => {
    const q = new URLSearchParams();
    if (params?.status) q.set("status", params.status);
    if (params?.limit) q.set("limit", String(params.limit));
    return req<SSCCPallet[]>("GET", `/sscc?${q}`);
  },
  getSSCC: (id: string) => req<SSCCPallet>("GET", `/sscc/${id}`),
  addLotToSSCC: (ssccId: string, lot_barcode_id: string, quantity?: number) =>
    req("POST", `/sscc/${ssccId}/lots`, { lot_barcode_id, quantity }),
  updateSSCCStatus: (id: string, status: string, shipment_reference?: string) => {
    const q = new URLSearchParams({ status });
    if (shipment_reference) q.set("shipment_reference", shipment_reference);
    return req("PATCH", `/sscc/${id}/status?${q}`);
  },

  // Label templates
  listTemplates: (packaging_level?: string) => {
    const q = packaging_level ? `?packaging_level=${packaging_level}` : "";
    return req<LabelTemplate[]>("GET", `/labels/templates${q}`);
  },
  getTemplate: (id: string) => req<LabelTemplate>("GET", `/labels/templates/${id}`),
  createTemplate: (d: Partial<LabelTemplate>) => req<LabelTemplate>("POST", "/labels/templates", d),
  updateTemplate: (id: string, d: Partial<LabelTemplate>) => req<LabelTemplate>("PATCH", `/labels/templates/${id}`, d),

  // Print jobs
  createPrintJob: (d: { trigger?: string; template_id?: string; items?: unknown[]; notes?: string }) =>
    req<PrintJob>("POST", "/labels/print", d),
  listPrintJobs: (status?: string) => req<PrintJob[]>("GET", `/labels/print${status ? `?status=${status}` : ""}`),
  completePrintJob: (id: string, printed_by = "user") =>
    req("POST", `/labels/print/${id}/complete?printed_by=${printed_by}`),

  // AI
  runLabelValidator: () => req<{ generated: number }>("POST", "/ai/run-label-validator"),
  runPackagingOptimizer: () => req<{ generated: number }>("POST", "/ai/run-packaging-optimizer"),
  listRecommendations: (agent_type?: string, status?: string) => {
    const q = new URLSearchParams();
    if (agent_type) q.set("agent_type", agent_type);
    if (status) q.set("status", status);
    return req<GS1AIRecommendation[]>("GET", `/ai/recommendations?${q}`);
  },
  reviewRecommendation: (id: string, status: string, reviewed_by?: string) =>
    req<GS1AIRecommendation>("PATCH", `/ai/recommendations/${id}`, { status, reviewed_by }),

  // Reports
  getPrintHistory: () => req<unknown[]>("GET", "/reports/print-history"),
  getSSCCTracking: () => req<unknown[]>("GET", "/reports/sscc-tracking"),
  getPackagingHierarchy: () => req<unknown[]>("GET", "/reports/packaging-hierarchy"),
  getBarcodeUsage: () => req<unknown[]>("GET", "/reports/barcode-usage"),
};

// ── Color Maps ────────────────────────────────────────────────────────────────

export const statusColor: Record<string, string> = {
  PENDING:    "bg-yellow-100 text-yellow-800",
  PRINTING:   "bg-blue-100 text-blue-800",
  COMPLETED:  "bg-green-100 text-green-800",
  FAILED:     "bg-red-100 text-red-800",
  CANCELLED:  "bg-gray-100 text-gray-600",
  ACTIVE:     "bg-green-100 text-green-800",
  SHIPPED:    "bg-blue-100 text-blue-800",
  RECEIVED:   "bg-purple-100 text-purple-800",
  DRAFT:      "bg-gray-100 text-gray-600",
  DEPRECATED: "bg-orange-100 text-orange-800",
};

export const severityColor: Record<string, string> = {
  CRITICAL: "bg-red-100 text-red-800",
  HIGH:     "bg-orange-100 text-orange-800",
  MEDIUM:   "bg-yellow-100 text-yellow-800",
  LOW:      "bg-green-100 text-green-800",
};

export const packagingLevelLabel: Record<PackagingLevel, string> = {
  UNIT:       "Unit / Each",
  INNER_PACK: "Inner Pack",
  CARTON:     "Carton",
  PALLET:     "Pallet",
};
